k2/csrc/fsa_algo.h

```diff
@@ -298,6 +298,38 @@ Ragged<int32_t> ShortestPath(FsaVec &fsas,
  */
 Fsa Union(FsaVec &fsas, Array1<int32_t> *arc_map = nullptr);
 
+/* Compute the closure of an FSA.
+
+   The closure is implemented in the following way:
+
+   - For all arcs entering the final-state, change the destination-state from
+   the final-state to state 0 (the start state), and the label from -1 to 0.
+
+   - Add an arc from the start-state to the final-state (with label -1, and
+   score 0, of course); this can be the last numbered arc of the start state of
+   the output. The implementation wouldn't be in-place, we'd create a new FSA.
+   Assume that the original start state has two arcs, then the arc-map would be
+   [ 0 1 -1 2 3 ... ].
```

Comment:
I just thought if we can implement it with such approach:
- add another super final_state, and let the original fina_state has an arc with label -1 to the super final state and an arc with label 0 to the start state
- change all final_arcs' labels from -1 to 0, still has the original final_state as the dest_state.

This approach may be simpler as all row_splits and row_ids in the original Fsa can be copied to the out FSA (as we don't insert an arc for start_state), all we need to do this to add one super final_state and two arcs for it. The logic in the lambda would be much simpler. Correct me if I'm wrong.

Reply:
I agree. The downside is we have one more state.

Reply:
If @danpovey  agrees with this approach, I would like to change the code.

Reply:
Another advantage of approach I metioned above is: if the input is top-sorted, then the ouput will be top-sorted.

Reply:
> I just thought maybe we can implement it with such approach:
> 
> * add another super final_state, and let the original fina_state has an arc with label -1 to the super final state and an arc with label 0 to the start state
> * change all final_arcs' labels from -1 to 0, still has the original final_state as the dest_state.
> 
> This approach may be simpler as all row_splits and row_ids in the original Fsa can be copied to the out FSA (as we don't insert an arc for start_state), all we need to do this to add one super final_state and two arcs for it. The logic in the lambda would be much simpler. Correct me if I'm wrong.

May still need to add an epsilon arc from the start state to the super final state to make the new fsa accept empty string

Reply:
Another advantage of the approach I metioned above is: we can make sure the output arc-sorted easily as the row_splits of the original FSA is not changed.  We can do `count` to know how many epsilon-arcs for each state, and then re-arrange the input arcs accoding to the number we just get. Just mention it, we may not need to implement arc-sort in this function.

Reply:
> May still need to add an epsilon arc from the start state to the original final state to make the new fsa accept empty string

Ah, OK, your requirement is A* instead of A+. Then we have to insert an arc for start state and can't use the approach I metioned above.